Disallow any compiler warnings by turning them into hard errors by the -Werror option.

AndersHoglund opened the pull request:
This is just an example and a suggestion for discussion how to block all compiler warnings and make them visible as hard red errors. I am not sure we want to be this strict. Currently all builds would be RED, mainly by some 11 warnings in the SITL target build.

ledvinap commented:
This was implemented in CF. Only PRs used -Werror, master did not fail on warning.

ledvinap commented:
https://github.com/cleanflight/cleanflight/blob/master_v1.x/.travis.sh#L24

AndersHoglund commented:
Oooh how I hate special cases. Different results on different branches with the same content.

unitware commented:
agree, its somewhat disturbing to rebase to the state of master only to find that your PR does not build because of some error introduced by some other forced PR merge. Such was the case in CF.

AndersHoglund commented:
Also, if PRs are clean then master will also be clean after a merge. Provided master was clean from the start, and that it was a trivial merge that did not add any new problems. And in this scenario there is no point at all to handle master and PRs differently.

AndersHoglund commented:
This is getting serious. PR #3066 introduced a new warning that slipped by un-noticed. We really need to implement this PR, forcing commits to be clean. But the baseline must be cleaned up, issue #3070 must be fixed first.

mikeller commented:
@AndersHoglund: 

#3066 is an example of a PR that would most likely never have happened if the (first time) contributor did not manage to get his change to build in the first place, which enabled him to test it, and then open the PR...

In this respect, if anything, we should change only the travis builds to break on warnings, enabling users to test their ideas without being pedantic about warnings.

AndersHoglund commented:
Building REVOLT with the #3066 changes locally would have shown the problem in the first place. Is that a common practice to either commit a PR with known warnings or not build locally at all? And then not checking CI results. OK there was no notification, this PR would fix that.

A user can test their ideas to any extent locally, having as much warnings and noise they ever would want. But committed code should be clean, at least to some degree. Minimum requirement would be that at least one target builds cleanly, especially if it is a target specific change.

Now it seems I have to modify my prev. statement:
"Oooh how I hate special cases. Different results in the local env vs in the CI env, or on different branches, with the same content."

mikeller commented:
@AndersHoglund: I am not saying that it should be common practice to push builds that introduce new warnings. In this case it was probably a user trying out a (very useful) patch, getting a warning, but then, when the patch actually worked, getting all excited about it and pushing the patch into a PR, warning at all. This is not ideal, but in my opinion still better than the user trying a new idea, then giving up because he doesn't manage to build because of warnings being turned into errors.

Another thing about warning is that, with 100+ targets, it is real easy to do a PR that works fine for the target that the user is working with, but causes warnings in other targets. Asking users to spend an hour or so for a `make all` before submitting a PR is just not realistic, so even turning warnings into errors will not help to reduce the number of buggy PRs that are opened.

What we really need is a CI system that is capable of properly reporting on warnings.

AndersHoglund commented:
I agree with you. But you misunderstand me a bit. A user should not have to `make all`, that is what CI should do. I said "...at least one target builds cleanly...". In this case minimum requirement would have been `make REVOLT` and check for issues.
Jenkins is now reporting Warnings as a yellow unstable result on slack,  however only on the master branch. 
Travis can not handle Warnings, unless they are turned into Errors by -Werror by this PR. In both cases the baseline MUST be clean and green. However, on the Jenkins site you can check the trend graph. If it goes above the current baseline of 11 warnings, Investigate!! Travis is still blind and green.

mikeller commented:
I think we want this:
https://www.jetbrains.com/buy/opensource/?product=teamcity

:-)

AndersHoglund commented:
Problem is not the tool itself, teamcity or jenkins or whatever. The problem is (free) hosting. This is where Travis wins. 
But please, feel free to setup a TeamCity  instance so we can try it out.

mikeller commented:
While it's not part of the problem, I think good tooling can be part of the solution. teamcity has the option to add more-or-less config free 'agents' to a server, and use the agents to run the builds. I think this model would enable us to distribute the load for building between multiple 'donors' of compute power, instead of having to rely on one provider to do all the work.

AndersHoglund commented:
Jenkins jobs can be distributed and use slave hosts too. Still a problem of finding one (free) hosts. Finding multiple is not any easier.
And whatever tool you have, a clean green baseline of the product code is a wanted state. Maybe even a requirement.

mikeller commented:
@AndersHoglund: Maybe having instructions on how users can set up slave hosts so that they can contribute processing power to our Jenkins setup would be a good start then. If we get enough people chipping in with slave hosts, we can move to running builds for PRs in addition to master.

(I just had a look at https://wiki.jenkins-ci.org/display/JENKINS/Step+by+step+guide+to+set+up+master+and+slave+machines+on+Windows , and the 'JWS' method seems to be zero config, and runnable from pretty much any workstation with a browser on it. That sounds like a good starting point.)

AndersHoglund commented:
Rebased to current master and removed the special case for SITL. All targets treated equal,  -Werror enabled on all. 
Now that Travis is updated and all targets actually compile nice and clean, we have the opportunity to turn on -Werror and catch future warnings as errors on Travis. Downside is that compiler and the whole make process will abort on first failure, if there are multiple warnings on multiple targets it could mean that those are not seen all at once and would perhaps be fixed one-by-one with multiple failing builds. 
But still I think the benefits with -Werror is worth that slightly annoying behaviour.

mikeller commented:
Can we change this so that the `-Werror` is only applied on travis builds? Having this on during development is more than annoying. (Potentially by adding a make option `RELEASE_BUILD`.)

AndersHoglund commented:
Sure. 
Added `EXTRAFLAGS` to the compiler command line. The `.travis.sh` now runs `make` with an argument `EXTRA_FLAGS=-Werror`. Can be done same way manually, if need. 
Also, this `EXTRA_FLAGS` opens up a possibility to try out any kind of compiler flag. The already available `OPTION` arg has a hardcoded `-D` prefix, making it limited to only add symbol definitions from the external env.
